### PR TITLE
diffex PDU support

### PR DESCRIPTION
**Pull request by bkmartinjr**
Initial Python and TypeScript modules supporting the encoding/decoding of the differential expression PDU.
* TS module in client/src/util/diffexpdu
* Py package in server/common/diffexpdu
* format spec in dev_docs/diffexpdu.md
* scripts to do Typescript/Python compatibility in scripts/diffexpdu 

Also includes basic unit tests.

### CODE REVIEW

**Comment by bkmartinjr** [comment]
Currently blocked by #230

File: server/common/diffexpdu/postingslist.py
Comment (blrnw3) [inline server/common/diffexpdu/postingslist.py:282]: I thought each block contains values from a single list? Indeed when decoding the blocks you have `assert len(list_ids) == 1`

File: server/common/diffexpdu/postingslist.py
Comment (blrnw3) [reply server/common/diffexpdu/postingslist.py:158]: I'm guessing this is to do with byte/delta shuffling which reduces the size by ~50%? If so what about starting with that uint encoding and falling back on bitarray if the reduction is not realized, i.e. the block is > 2^16 bits?

**Review by blrnw3 — COMMENTED** [review]
I focussed on the python implementation since that's easier for me to follow. Let me know if you want a deeper review on any part of the ts code.
Anyway aside from my questions it's looking good to me, plenty readable and thanks for all the docs!

File: dev_docs/diffexpdu.md
Comment (bkmartinjr) [reply dev_docs/diffexpdu.md:83]: fixed

File: dev_docs/diffexpdu.md
Comment (bkmartinjr) [reply dev_docs/diffexpdu.md:65]: I left open the possibility of testing for inverted lists (in other block types).  However, I don't see any value at this point, and will revert this.

File: dev_docs/diffexpdu.md
Comment (bkmartinjr) [reply dev_docs/diffexpdu.md:61]: yes.  Will attempt to clarify.

File: dev_docs/diffexpdu.md
Comment (bkmartinjr) [reply dev_docs/diffexpdu.md:69]: will do

File: server/common/diffexpdu/diffexpdu.py
Comment (bkmartinjr) [reply server/common/diffexpdu/diffexpdu.py:24]: fixed

File: dev_docs/diffexpdu.md
Comment (bkmartinjr) [reply dev_docs/diffexpdu.md:1]: https://en.wikipedia.org/wiki/Protocol_data_unit

Very commonly used term for this type of network encoding data structure.

I'm open to changing any of the names for clarity, so please suggest alternatives.

File: server/tests/unit/common/test_diffexpdu.py
Comment (bkmartinjr) [reply server/tests/unit/common/test_diffexpdu.py:49]: We could, but I didn't expose public API for this (for obvious reasons that I didn't want anyone to code to non-public interfaces).

How do you feel about tests that utilize internal API?  Very easy to add and test, as long as we recognize that the tests will likely have to change when we evolve the format.

File: server/common/diffexpdu/postingslist.py
Comment (bkmartinjr) [reply server/common/diffexpdu/postingslist.py:36]: bug.  fixed!

Comment (bkmartinjr) [reply server/common/diffexpdu/postingslist.py:36]: bug.  fixed!

Comment (bkmartinjr) [reply server/common/diffexpdu/postingslist.py:36]: bug, fixed!

File: server/common/diffexpdu/postingslist.py
Comment (bkmartinjr) [reply server/common/diffexpdu/postingslist.py:100]: fixed

File: server/common/diffexpdu/postingslist.py
Comment (bkmartinjr) [reply server/common/diffexpdu/postingslist.py:158]: It primarily has to do with the subsequent deflate compression performance (deflate is applied to all block types).  The amount of compression is highly dependent on the actual list values, not their encoding.  While it is super easy to predict the cut-off for uncompressed blocks, it is not easy to do it for compressed blocks (without calling deflate()).

And the reason I didn't want to call deflate() is that it is expensive - doing the prospective compression in all formats, and then picking the smallest, would at least double the cost of encoding a block.  For large cell sets, this gets prohibitive on the front-end.

So, what I did is generate a very large set of test cases (over 75K) from real data, and empirically determine a reasonable cut-off.  Turns out that the heuristic appears reasonable from first principles (eg, small sets are better to encode with a discrete list,  very large lists are better to code with an inverted list, and all others are better as a bitarray).  The density clamp turns out to control for a very rough estimate on entropy, further refining the heuristic guess of how well deflate will work.

I'm happy to share the code I used to derive these specific values if you are interested.  I didn't think it was worth putting in the repo as it is not a runtime dependency (but LMK if you disagree).


As an aside -- the reason more exotic block types (eg, multi-list disjoint encodings, etc) did not provide significant benefits is because we use deflate on all block types.

File: server/common/diffexpdu/postingslist.py
Comment (bkmartinjr) [reply server/common/diffexpdu/postingslist.py:282]: True, for now.  This extra feature is in place to allow multi-list blocks if we find they have value.  I prototyped it, and did not find a big win, so I left out the actual block type (but retained the framework).

Currently, any single BitArray or Uint16 block can contain a single list - the list is identified by the "list ID mask" which can contain only a single bit set.   A diffexp PDU will have blocks from multiple lists, with their specific list ID encoded in this mask (currently diffex uses 2 lists, but the format supports up to 8 as we may find other uses for it).

We could simplify this slightly by changing the list "mask" to a list "id", but there is no major value in doing so, and it would remove the ability to add multi-list blocks in the future.

**Comment by bkmartinjr** [comment]
@blrnw3 - I also realized that I neglected to add a header byte to the diffexpdu format, which I think is important for versioning (ie, if we want to change the format in the future).  I have one already embedded in the postings list in case we add a block type, etc.

Will push a few more commits to add this top-level byte.  Documenting rationale here in case you have concerns.

File: server/tests/unit/common/test_diffexpdu.py
Comment (bkmartinjr) [reply server/tests/unit/common/test_diffexpdu.py:49]: BTW, I am getting pretty good coverage on the tests, which I think is a positive indication:
```
File                                   | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s  
---------------------------------------|---------|----------|---------|---------|--------------------
 src/util/diffexpdu                    |   95.51 |    77.27 |   95.65 |   95.86 | 
  diffexpdu.ts                         |   86.36 |     62.5 |     100 |   86.36 | 39,41,83 
  index.ts                             |       0 |        0 |       0 |       0 | 
  pdubuffer.ts                         |   88.46 |       60 |   88.89 |      88 | 20,56-57    
  postingslist.ts                      |   96.94 |       80 |     100 |   97.61 | 47,50,228,278,286  
  utils.ts                             |   97.14 |     87.5 |   85.71 |   97.06 | 41  
 ```

Comment (blrnw3) [reply server/tests/unit/common/test_diffexpdu.py:49]: This makes sense, thanks for the coverage report. I usually prefer not testing the internal APIs unless there are specific outcomes to verify. In this case I don't have a strong preference so will defer to yours.

File: server/common/diffexpdu/postingslist.py
Comment (blrnw3) [reply server/common/diffexpdu/postingslist.py:157]: Thanks for this context. This all makes sense. I think it would be helpful to have a version of this explanation in the code if possible, to aid future readers.

Comment (bkmartinjr) [reply server/common/diffexpdu/postingslist.py:157]: both the Py and TS code has the same heuristic, so I'll add a comment there pointing to the dev_docs, and add a section there describing the tradeoffs/heuristic.

File: server/common/diffexpdu/postingslist.py
Comment (bkmartinjr) [reply server/common/diffexpdu/postingslist.py:287]: I added a note about this to the docs as well.

File: server/tests/unit/common/test_diffexpdu.py
Comment (blrnw3) [reply server/tests/unit/common/test_diffexpdu.py:49]: Ha for sure. I'm good with where you landed

**Comment by blrnw3** [comment]
Thanks for the latest changes. I'm looking into the dependencies bug which is blocking you now.

**Review by blrnw3 — APPROVED** [review]
Thanks Bruce! Looks great to me

**Pull request merged**
